### ISSUE TITLE
System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows test failure

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#76831]
```
   System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows [FAIL]
      System.AggregateException : One or more errors occurred. (Assert.Throws() Failure
      Expected: typeof(System.Net.Quic.QuicException)
      Actual:   (No exception was thrown)) (Assert.Throws() Failure
      Expected: typeof(System.Net.Quic.QuicException)
      Actual:   (No exception was thrown))
      ---- Assert.Throws() Failure
      Expected: typeof(System.Net.Quic.QuicException)
      Actual:   (No exception was thrown)
      ---- Assert.Throws() Failure
      Expected: typeof(System.Net.Quic.QuicException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(92,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks)
        /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(55,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(303,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(303,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 
```

Failed in #76768. Full log: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-76768-merge-80851214628f4109bd/System.Net.Quic.Functional.Tests/1/console.21da1d39.log?helixlogtype=result

```json
{
  "ErrorMessage": "System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows",
  "BuildRetry": false
}
```

### Report
|Build|Definition|Step Name|Console log|Pull Request|
|---|---|---|---|---|
|[392518](https://dev.azure.com/dnceng-public/public/_build/results?buildId=392518)|dotnet/runtime|Send to Helix|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/392518/logs/815)||
|[388991](https://dev.azure.com/dnceng-public/public/_build/results?buildId=388991)|dotnet/runtime|Send to Helix|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/388991/logs/852)||
|[384892](https://dev.azure.com/dnceng-public/public/_build/results?buildId=384892)|dotnet/runtime|Send to Helix|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/384892/logs/825)||

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[411093](https://dev.azure.com/dnceng-public/public/_build/results?buildId=411093)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.ReadWrite_Random_Success](https://dev.azure.com/dnceng-public/public/_build/results?buildId=411093&view=ms.vss-test-web.build-test-results-tab&runId=8947652&resultId=169002)||
|[410091](https://dev.azure.com/dnceng-public/public/_build/results?buildId=410091)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=410091&view=ms.vss-test-web.build-test-results-tab&runId=8919054&resultId=200435)||
|[408592](https://dev.azure.com/dnceng-public/public/_build/results?buildId=408592)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.ReadWrite_Random_Success](https://dev.azure.com/dnceng-public/public/_build/results?buildId=408592&view=ms.vss-test-web.build-test-results-tab&runId=8874378&resultId=173366)||
|[408334](https://dev.azure.com/dnceng-public/public/_build/results?buildId=408334)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=408334&view=ms.vss-test-web.build-test-results-tab&runId=8867008&resultId=200541)||
|[407850](https://dev.azure.com/dnceng-public/public/_build/results?buildId=407850)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=407850&view=ms.vss-test-web.build-test-results-tab&runId=8853878&resultId=200540)||
|[406522](https://dev.azure.com/dnceng-public/public/_build/results?buildId=406522)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=406522&view=ms.vss-test-web.build-test-results-tab&runId=8821960&resultId=200407)||
|[406371](https://dev.azure.com/dnceng-public/public/_build/results?buildId=406371)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamConformanceTests.ReadWrite_Success](https://dev.azure.com/dnceng-public/public/_build/results?buildId=406371&view=ms.vss-test-web.build-test-results-tab&runId=8818154&resultId=162026)|dotnet/runtime#92026|
|[401264](https://dev.azure.com/dnceng-public/public/_build/results?buildId=401264)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=401264&view=ms.vss-test-web.build-test-results-tab&runId=8670392&resultId=200449)||
|[401214](https://dev.azure.com/dnceng-public/public/_build/results?buildId=401214)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=401214&view=ms.vss-test-web.build-test-results-tab&runId=8668880&resultId=200565)||
|[397203](https://dev.azure.com/dnceng-public/public/_build/results?buildId=397203)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=397203&view=ms.vss-test-web.build-test-results-tab&runId=8572860&resultId=200432)|dotnet/runtime#91683|
|[396223](https://dev.azure.com/dnceng-public/public/_build/results?buildId=396223)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=396223&view=ms.vss-test-web.build-test-results-tab&runId=8547054&resultId=194739)|dotnet/runtime#91140|
|[393529](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393529)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393529&view=ms.vss-test-web.build-test-results-tab&runId=8465548&resultId=195961)||
|[393353](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393353)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393353&view=ms.vss-test-web.build-test-results-tab&runId=8461362&resultId=200448)|dotnet/runtime#91468|
|[393144](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393144)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=393144&view=ms.vss-test-web.build-test-results-tab&runId=8455214&resultId=194755)|dotnet/runtime#91413|
|[392801](https://dev.azure.com/dnceng-public/public/_build/results?buildId=392801)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=392801&view=ms.vss-test-web.build-test-results-tab&runId=8445402&resultId=200417)|dotnet/runtime#91451|
|[392687](https://dev.azure.com/dnceng-public/public/_build/results?buildId=392687)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=392687&view=ms.vss-test-web.build-test-results-tab&runId=8441338&resultId=195623)||
|[389670](https://dev.azure.com/dnceng-public/public/_build/results?buildId=389670)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=389670&view=ms.vss-test-web.build-test-results-tab&runId=8353968&resultId=200407)||
|[388225](https://dev.azure.com/dnceng-public/public/_build/results?buildId=388225)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=388225&view=ms.vss-test-web.build-test-results-tab&runId=8317434&resultId=200404)|dotnet/runtime#91224|
|[387254](https://dev.azure.com/dnceng-public/public/_build/results?buildId=387254)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=387254&view=ms.vss-test-web.build-test-results-tab&runId=8292042&resultId=194353)|dotnet/runtime#91166|
|[386741](https://dev.azure.com/dnceng-public/public/_build/results?buildId=386741)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=386741&view=ms.vss-test-web.build-test-results-tab&runId=8272328&resultId=200384)||
|[381472](https://dev.azure.com/dnceng-public/public/_build/results?buildId=381472)|dotnet/runtime|[System.Net.Quic.Tests.QuicStreamTests.WriteCanceled_NextWriteThrows](https://dev.azure.com/dnceng-public/public/_build/results?buildId=381472&view=ms.vss-test-web.build-test-results-tab&runId=8135264&resultId=200086)|dotnet/runtime#90794|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|2|7|24|

 ### Known issue validation
**Build: :mag_right:** 
**Result validation: :warning:** Validation could not be done without an Azure DevOps build URL on the issue. Please add it to the "**Build: :mag_right:**" line.

**Comment by rzikm** [comment]
The test is flaky and the failure happens on regular runs as well (but usually succeeds on reruns)

The case seems to be when the cancellation happens to occur before the call to WriteAsync (and we call it with pre-cancelled token).

https://github.com/dotnet/runtime/blob/844702d736eef2fc456c1e4d89043fab6d054609/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs#L591-L606

We should have a look someday.

**Comment by rzikm** [comment]
Okay, my original thought was wrong, we handle precancelled tokens correctly, there is actually very slim chance to repro this because many things have to align properly.

the issue can be artificially enforced reproduced by adding sleep in
- Abort (into `if (abortDirection.HasFlag(QuicAbortDirection.Write))` branch)
- WriteAsync (after `if (!_sendTcs.TryGetValueTask(out ValueTask valueTask, this, cancellationToken))`) or by making huge writes (e.g. 1 GB) to ensure that CancellationTokenSource fires while send is in progress

What happens is that cancellation token sets Write state to cancelled and unblocks the Test task to progress, and it completes the `clientFunction` and `Dispose`s the client `QuicStream` (i.e. completing the write direction gracefully), **before** the cancellation-registered action (which runs on a different thread) manages to finish the `Abort` call.

In other words, if cancellation token fires while send is in progress, we may end up gracefully closing the write side :/
